Code/Framework/AzCore/AzCore/Jobs/Internal/JobTypeEraser.h

```diff
@@ -0,0 +1,232 @@
+    //
+    // Lambdas that are trivially destructible will result in a nullptr returned JobDestroy_t pointer.
+    //
+    // The class will check that the lambda is copy assignable or movable.
+    template<typename Lambda>
+    class JobTypeEraser final
+    {
+    public:
+        constexpr JobInvoke_t ErasedInvoker()
+        {
+            return reinterpret_cast<JobInvoke_t>(Invoker);
+        }
+
+        constexpr JobRelocate_t ErasedRelocator()
+        {
+            if constexpr (AZStd::is_trivially_move_constructible_v<Lambda>)
+            {
+                return nullptr;
+            }
+            else if constexpr (AZStd::is_move_constructible_v<Lambda>)
+            {
+                return reinterpret_cast<JobRelocate_t>(Mover);
+            }
+            else if constexpr (AZStd::is_copy_constructible_v<Lambda>)
+            {
+                return reinterpret_cast<JobRelocate_t>(Copyer);
+            }
+            else
+            {
+                static_assert(
+                    AZStd::is_move_constructible_v<Lambda> || AZStd::is_copy_constructible_v<Lambda>,
+                    "Job lambdas must be either move or copy constructible. Please verify that all captured data is move or copy "
+                    "constructible.");
+            }
+        }
+
+        constexpr JobDestroy_t ErasedDestroyer()
+        {
+            if constexpr (AZStd::is_trivially_destructible_v<Lambda>)
+            {
+                return nullptr;
+            }
+            else
+            {
+                return reinterpret_cast<JobDestroy_t>(Destroyer);
+            }
+        }
+
+    private:
+        constexpr static void Invoker(Lambda* lambda)
+        {
+            lambda->operator()();
+        }
+
+        constexpr static void Mover(Lambda* dst, Lambda* src)
+        {
+            new (dst) Lambda{ AZStd::move(*src) };
+        }
+
+        constexpr static void Copyer(Lambda* dst, Lambda* src)
+        {
+            new (dst) Lambda{ *src };
+        }
+
+        constexpr static void Destroyer(Lambda* lambda)
+        {
+            lambda->~Lambda();
+        }
+    };
+
+    // The TypeErasedJob encapsulates member function pointers to store in a homogeneously-typed container
+    // The function signature of all lambdas encoded in a TypeErasedJob is void(*)(). The lambdas can capture
+    // data, in which case the data is inlined in this structure if the payload is less than or equal to the
+    // buffer size. Otherwise, the data is heap allocated.
+    class alignas(alignof(max_align_t)) TypeErasedJob final
+    {
+    public:
+        // The inline buffer allows the TypeErasedJob to span two cache lines. Lambdas can capture 48
+        // bytes of data (6 pointers/references on a 64-bit machine) before spilling to the heap.
+        constexpr static size_t BufferSize =
+            128 - sizeof(size_t) * 6 - sizeof(uint32_t) - sizeof(JobDescriptor) - sizeof(AZStd::atomic<uint32_t>);
```

Comment:
Yup

Reply:
While this is in flux, I'm going to leave config in a new `TaskConfig.h` file until we have a need to separate it into per-platform traits (just to keep iteration time fast since this is a new system)